### ISSUE TITLE
Manual Configuration: multiple environments

### ISSUE THREAD

**Issue by mageddo** [mageddo/dns-proxy-server#76]
**Context**
When you use DPS local storage you can register many hostname records referring a custom IP, that way you can use hostnames instead of static IPs then change the IPs when it comes necessary in one place. It have nothing with docker, just DPS working like a DNS with advanced settings

**Use case Proposal**
It would be nice if you can register many environments so you can register multiple IPs for the same hostname then you can point to a different IP by switching the environment.

Example 

Let's say you own **acme.com** and is coding on that, you probably want to access acme.com pointing to your local machine instead of production, so you would use **Dev** environment that has a `127.0.0.1` entry at **acme.com**.

If you are testing QA then you swap to **QA** environment.

Then to default/prod when want  to see it in prod, this environment probably have no entries so DPS will solve to the real IP

**Requirements**
* Using GUI user can create/delete/update as many environments as wanted
* User can choose what the active environment
* Every environment has it's own hostnames entries

**Comment by josielrocha** [comment]
Hello @mageddo!

I've forked your project (https://github.com/josielrocha/dns-proxy-server) and I'm trying to implement this feature.. I found something that I need to know from you about how to deal with it.

Step by step:
1. Create a new Environment (Test)
2. Add a host, (acme.com -> 127.0.0.1)
3. Switch back to default environment
4. Add a host, (acme.com -> 127.0.0.2)
5. Query the hostname in the **Test** environment (dig acme.com => 127.0.0.1)
6. Switch back to default envrionment
7. Query the hostname in the **default** environment (dig acme.com => 127.0.0.1, I was expecting 127.0.0.2)

It seems to be caching the query result without respecting the environment.. It is like the cache key is composed only by hostname

**UI with Test environment and acme.com host**
![image](https://user-images.githubusercontent.com/1158643/60751295-7db72000-9f89-11e9-8c9b-9ee21396140f.png)

**UI with Default environment and acme.com host**
![image](https://user-images.githubusercontent.com/1158643/60751314-a93a0a80-9f89-11e9-9838-d2d073b5078c.png)

```
$ curl -Ri http://localhost:5380/env/active
HTTP/1.1 200 OK
Content-Type: application/json
Date: Sat, 06 Jul 2019 04:03:35 GMT
Content-Length: 20

{
	"name": "Test"
}
$ dig acme.com

; <<>> DiG 9.14.3 <<>> acme.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 26209
;; flags: qr rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;acme.com.			IN	A

;; ANSWER SECTION:
acme.com.		60	IN	A	127.0.0.2  <---- IP from Test environment

;; Query time: 1 msec
;; SERVER: 192.168.1.50#53(192.168.1.50)
;; WHEN: Sat Jul 06 01:03:41 -03 2019
;; MSG SIZE  rcvd: 50

$ curl -Ri http://localhost:5380/env/active
HTTP/1.1 200 OK
Content-Type: application/json
Date: Sat, 06 Jul 2019 04:03:54 GMT
Content-Length: 16

{
	"name": ""
}
$ dig acme.com

; <<>> DiG 9.14.3 <<>> acme.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 749
;; flags: qr rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;acme.com.			IN	A

;; ANSWER SECTION:
acme.com.		45	IN	A	127.0.0.2  <------ Same IP as Test environment

;; Query time: 2 msec
;; SERVER: 192.168.1.50#53(192.168.1.50)
;; WHEN: Sat Jul 06 01:03:56 -03 2019
;; MSG SIZE  rcvd: 50
```
Is it expected? Do I need to clear caches after switching environments?

**Comment by mageddo** [comment]
> Is it expected? Do I need to clear caches after switching environments?

Yep, take a look at the file which read the json configuration file,it caches the entire file, you will have to clear the cache

**Comment by mageddo** [comment]
Are you calling the method https://github.com/mageddo/dns-proxy-server/blob/master/events/local/local.go#L172 when setting the active env? It should so the trick cause it store the changes to the file and also clear the cache

**Comment by josielrocha** [comment]
The REST API is working perfectly fine.. The problem occurs when we send queries to the Dns Proxy Server..

I was looking the code and I believe we need to pass active environment to Local and Cache solvers in this line (https://github.com/mageddo/dns-proxy-server/blob/master/dns.go#L74) or let Cache and Local solvers know what is the current environment.

Do you agree with my point?

**Comment by mageddo** [comment]
DNS resolution is working fine, I tested it, you have to change active env first 

```
curl -X PUT \
  http://dps.docker:5380/env/active \
  -H 'Content-Type: application/json' \
  -d '{ "name": "myEnv" }'
```

**Comment by mageddo** [comment]
Some changes will be necessary on the webpage api clients though, they must pass the env where the hostname will be created, deleted, etc.

**Comment by mageddo** [comment]
> The problem occurs when we send queries to the Dns Proxy Server..

Yep, DPS is really having issues when the environment is switched, I think the cache must be clearedm I'm working on it

**Comment by mageddo** [comment]
#155 created to fix env switch resolution issues

**Comment by mageddo** [comment]
@josielrocha I've forked your branch, made some minor changes then merged it #154

**Comment by mageddo** [comment]
@josielrocha [here](http://mageddo.github.io/dns-proxy-server/latest/en/credits/) your credits
